Index: social_validator/twitch.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/social_validator/twitch.py b/social_validator/twitch.py
new file mode 100644
--- /dev/null	(date 1699414637088)
+++ b/social_validator/twitch.py	(date 1699414637088)
@@ -0,0 +1,20 @@
+from social_validator import shared
+
+# Restrictions for usernames
+USERNAME_MIN_LENGTH = 4
+USERNAME_MAX_LENGTH = 25
+
+
+def _is_valid_first_char(s: str) -> bool:
+    if not s:  # empty string
+        return False
+
+    return s[0] != "_"
+
+
+def is_valid_username(username: str) -> bool:
+    return (
+            _is_valid_first_char(username)
+            and USERNAME_MIN_LENGTH <= len(username) <= USERNAME_MAX_LENGTH
+            and shared.is_valid_id(username)
+    )
